rfcs/006-package-management.md

```diff
@@ -0,0 +1,560 @@
+
+### Alternative: allow yanked packages if resolution would otherwise fail
+
+A slightly more permissive alternative would be to ignore yanked packages that
+aren't in the lock-file, but if resolution fails then go back and try allowing
+them. This would allow resolution to succeed, for example, if your manifest
+asks for the latest version of a package *and* that package was yanked *and*
+you don't have a lock-file. But this seems like maybe adding too many corner-cases.
+
+## The registry
+
+How should we manage the global registry? There's a potential for incurring
+substantial maintenance costs here, so we should be careful.
+
+We will provide a git repo, hard-coded to live at `github.com/nickel-lang/nickel-mine`,
+to serve as the registry.
+This repo will contain the "index", but not the actual package contents. It
+will contain one file per package, each of which contains a line per version.
+Each entry specifies the location of the package (currently required to be on
+github) and its git tree hash. This ensures that packages are immutable, but it
+doesn't stop them from disappearing: we don't keep a copy of the actual package
+contents.
+
+The registry index will also contain package metadata: the dependencies (so
+that we can resolve recursive dependencies without fetching the packages),
+and whatever metadata we want to be searchable.
+
+The registry entries are named like "github/\<org\>/\<package\>" (where in the
+future we might support places other than github). This allows us to skip
+registration and authentication: if someone has the github permissions to create
+`jneem/foo` on github, they also have permissions to create and update the
+`github/jneem/foo` nickel package.
+
+Packages will be updated by pull requests to the index. There can be some tooling
+in the nickel CLI to automate the creation of these pull requests. If the volume
+of updates becomes high enough, we can automate the merging.
+
+### Question: should we store a content hash too?
+
+We're storing a git tree hash in the index, but if we ever want to store package
+contents in the future, maybe we should also store a hash of the git tree
+contents? This would allow verification of package tarballs, without needing the
+whole git repo.
+
+## Alternate registries
+
+It's very convenient to have a default global registry, but you might also want
+to get the benefits of a registry without making your code public. Therefore
+we should support alternatives to the global registry. There are at least two
+distinct use-cases:
+
+1. replacing the default registry with an alternative, for example in order to
+   host or cache it locally
+2. providing an additional registry with private code, to be used alongside
+   the global registry
+
+These two use-cases should be configured differently, because the first is
+a global setting and the second is per-package. For the global setting, we
+add a `source-replacement` field to the manifest: `std.package.Manifest` becomes
```

Comment:
I think I don't understand: you say it's a global setting, but we still configure it at the level of the manifest of a package? Or by global do you mean it applies to all the dependencies of one package,  instead of potentially one of the  dependency of one package?